### DISCUSSION TITLE
Requesting feedback on a relevant jsPsych library

### DISCUSSION THREAD

**Discussion by vijaymarupudi** [jspsych/jsPsych#732] [Q&A]
[Cross-posting on Github]

Hello people interested in jsPsych,

For the past month I've been helping out people with their jsPsych problems, and have noticed several patterns in the errors that happen. People sometimes want to do something complicated involving a dynamic number of trials. People want to evaluate certain parameters at the time of the experiment rather than before it begins.

I believe a lot of this stems from a lack of understanding of the asynchronous nature of Javascript and the complexity of lifecycle callbacks. When jsPsych was first made, back in 2012, I believe there was no other way to deal with these problems. However the Javascript language and ecosystem have evolved in that time, and I'd like to incorporate those ideas to make designing online experiments easier.

I decided to write a library to do so. I was heavily inspired by jsPsych's elegance, and will continue to draw on it for code and ideas. It is very much in progress, but is currently functional. My primary goal was to eliminate the common questions by designing a different API. This required a department from the declarative nature of jsPsych towards a more runtime specified design. The Javascript feature I have focused on using are [generator functions]. They allow executation of code to be paused and played at specific points of the function.

[generator functions]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*

Generator functions allow for a more intuitive flow of code that I think is easier for a beginner, a common user of jsPsych, to understand. Here is an example of a simple experiment that requires participants to click a button within 500ms of it being displayed. If they do not succeed, they would have to try again. In jsPsych, this would require a loop_function and an on_finish callback, however here it is in the library I have made.

Make sure you have jsPsych's js, css, and the appropriate plugin js loaded.

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <script src="jsPsych/jspsych.js"></script>
    <script src="jsPsych/plugins/jspsych-html-button-response.js"></script>
    <script src="dist/reaction-time.js"></script>
    <link rel="stylesheet" href="jsPsych/css/jspsych.css" />
  </head>
  <body>
    <script>
      function* timelineGenerator(reactionTime) {

        while (true) {
          const result = yield ReactionTime.jsPsychPlugin({
            type: "html-button-response",
            stimulus: "<p>Click fast</p>",
            choices: ["Here"]
          });

          console.log(result); // data for that specific trial
          console.log(Date.now()); // timestamp for the moment the trial is over

          if (result.rt < 500) {
            break;
          }

          yield ReactionTime.jsPsychPlugin({
            type: "html-button-response",
            stimulus: "<p>Too slow! Your RT was " + result.rt.toString() + "</p>",
            choices: ["Try again"]
          })
        }

        reactionTime.displayData();

      }

      ReactionTime.init(timelineGenerator);
    </script>
  </body>
</html>
```

[I have attached a fully working sample with the library code in this zip file.](https://send.firefox.com/download/0d096ecb984fd12e/#cWWC2anx6cHFYXjo6gRucQ) Click `index.html` to enjoy! If you'd like to explore the library code, check out the `src/ReactionTime.ts`, `src/reaction-time.ts`, and the `src/jspsych-plugin.ts` files.  

Source code is also available at <https://github.com/vijaymarupudi/reaction-time>

I have named the library `ReactionTime` for now, hopefully that is not too confusing. `reactionTime` is an interface to the library. I'd like opinions on what people think of this API. Also you may have noticed that I was using a jsPsych plugin! While the plugin system is a little different for my library, they can are compatible with each other.

The code for this API is surprisingly simple and short. I have wrapper almost all jsPsych plugins (and most of jsPsych) within 100 lines of code. The library is written in Typescript for easier long term maintenance and useful editor hints in editors like Atom, VSCode, and Vim. I make liberal use of modern Javascript features (Promises & Async functions) knowing that I can transpile them to work in old browsers. Features like `Proxy` that aren't possible in older browsers have been avoided.

A few notes on design philosophy. I have a design goal to keep this library simple and easy to maintain. I personally think that tasks that can be trivially performed in a loop should be done using loops. That would mean no timeline variables and no looping or conditional functions. I do realize that the declarative API that jsPsych provides makes it very easy to design simple experiments, I think the lack of an "escape hatch" makes it difficult to use for complicated ones that I have seen in the wild.

I think this api would make the follow features mentioned in the documentation redundant because generator functions make asynchronous code appear synchronous: 

* Functions for stimulus values (dynamic parameters).
* Repeating trials
* Looping timelines
* Conditional timelines
* data parameter
* Trial lifecycle callbacks

One con of this approach is that autopreloading is impossible. I think that's an acceptable tradeoff and that is it okay to ask the user to call the preload function with their list of media.

Aesthetically, I have completely avoided globals (such as depending on a single display_element being defined for the whole page) for the plugin API. I also plan on defering a lot of the work that's in `jspsych.js` to the plugins (specifically multimedia associated code). Given the modular design of the code, this would make it easier to maintain different plugins of the same media type.

Wrapping jsPsych was an important goal for this library, as the work determining the nuances and quirks of browsers and multimedia playback is all stored in the code and I would not want to rewrite all of it. I hope for new plugins to be written using the `React` library with reusable hooks to maintain side-effects and state, but whether that will happen will depend on intersect.

For future directions, I'd like to see if there's any interest at all. I plan on using this as a personal library to design some of the more complicated experiments, but if there's interest there are many options. One option would be to make this API the 'escape-hatch' in jsPsych ala `jsPsych.init_generator()`. Another not mutually exclusive option would be to publish it and maintain it alongside jsPsych.

I think a lot of the value in jsPsych and probably the time spent on it was the documentation and tutorials. So if there's interest, I might spend some time on this API. Let me know your thoughts!

License: LGPL-3.0 or later

\~ Vijay

**Comment by jodeleeuw** [comment]
Hi @vijaymarupudi ! Sorry it has taken me a couple of days to respond to this. I needed to make sure I had enough time to think about it and write a reply.

Here are my reactions, in no particular order:

1. I really like the idea of providing an alternative method for generating a timeline via aysnc functions, especially if you've found a way to elegantly wrap plugins so that they can work in either context. I even wonder if timelines could be hybrids of declarative and generative with the right API design.
2. I'm not sure whether the declarative or generative model will be easier for new users. I think that's an argument for making two options available. The risk of making two options available is that it's less clear where to go when getting started. But I think a good set of documentation and tutorials could clarify this.
3. I agree that this mode of writing experiments would make a lot of the features you identify redundant. But I think that sometimes redundancy can also be useful abstraction. For example, timeline variables didn't exist in early versions of jsPsych, but one could implement the same experiment through the use of arrays and functions as parameters. 
4. Could you elaborate on how you are handling the `display_element`? I'm not sure I follow. It's a global in the sense that it is scoped to the `jsPsych` root object, 
5. You mention deferring work to the plugins instead of jspsych.js. That's interesting, because I've been inclined to go the other direction recently! I've been thinking about creating code templates for different elements that exist centrally, and allowing plugins to use these when desired. My thought is this makes it easier to maintain plugin elements that are reused across multiple plugins. It also makes it easier to design new plugins by recombining existing components. What do you think the advantages are of decentralizing? What pieces of code were you thinking about moving?
6. I'm *really* hesitant to make plugins depend on `React`. In general I want to avoid build chains for jsPsych, because I think it creates a significant barrier for new users. Some of my comments [here](https://github.com/jspsych/jsPsych/issues/625) are relevant. What do you think the major advantages of adopting React would be? Are those advantages things that we can get with smaller changes to the plugin format?

**Reply by vijaymarupudi** [reply]
I appreciate your thoughts, thank you! I plan on playing with this idea a bit further and seeing where it goes.

> I really like the idea of providing an alternative method for generating a timeline via async functions, especially if you've found a way to elegantly wrap plugins so that they can work in either context. I even wonder if timelines could be hybrids of declarative and generative with the right API design.

I don't fully understand jsPsych's timeline node system, in fact that was one of the motivations for me to write it this way so that the Javascript stack takes care of the nesting. I assume you cannot dynamically add trials anywhere between the timeline, which this API would allow you to do, so maybe they are incompatible?

> I'm not sure whether the declarative or generative model will be easier for new users. I think that's an argument for making two options available. The risk of making two options available is that it's less clear where to go when getting started. But I think a good set of documentation and tutorials could clarify this.

I think the current declarative system is better for simple experiments i.e. the user does not need to care too much about looping and randomization. However I think when you have logic before and after trials and different behavior as a result of it, then the declarative API becomes complex very fast. While this approach might be slightly more complex to start, I think the composability it provides will be easier as they continue to use it. A lot of my experiments in math cognition / categorization would benefit from this API.

> I agree that this mode of writing experiments would make a lot of the features you identify redundant. But I think that sometimes redundancy can also be useful abstraction. For example, timeline variables didn't exist in early versions of jsPsych, but one could implement the same experiment through the use of arrays and functions as parameters.

I think that's fair, and the features do provide useful abstractions! My comment was more towards how the abstractions compose or become more difficult as the complexity of the design increses, an example being needing to use `true` as the second argument to `jsPsych.timelineVariable()` when in a function.

> Could you elaborate on how you are handling the display_element? I'm not sure I follow. It's a global in the sense that it is scoped to the jsPsych root object

It's global as in there's only one, maybe singleton is a better choice of word. I am thinking of 'viewing' plugins, that display stimuli, and 'response' plugins, that deal with user input. Both plugins need a `display_element`, but deal with them differently. That way the user can compose between the type of response they'd like and the stimuli they'd like to show.

> You mention deferring work to the plugins instead of jspsych.js. That's interesting, because I've been inclined to go the other direction recently! I've been thinking about creating code templates for different elements that exist centrally, and allowing plugins to use these when desired. My thought is this makes it easier to maintain plugin elements that are reused across multiple plugins. It also makes it easier to design new plugins by recombining existing components. What do you think the advantages are of decentralizing? What pieces of code were you thinking about moving?

My opinion was similar, but instead of moving the common code to the core, the code would be moved to a utility library. I think if the core is kept simple and independent of the plugins, it would be easier to maintain. This is me taking what I find is the elegance of jsPsych to the extreme.

> I'm really hesitant to make plugins depend on React. In general I want to avoid build chains for jsPsych, because I think it creates a significant barrier for new users. Some of my comments here are relevant. What do you think the major advantages of adopting React would be? Are those advantages things that we can get with smaller changes to the plugin format?

You can make it so that the users who don't care about the build chain don't need to, just provide them a transpiled bundle to use, kinda like how you can do so with React (without JSX of course). Those who want to use the library using npm can do so using the es module version. Given React's license, one can bundle it along with the library code, making it seamless for the user to use.

I think the primary advantage to using React would be to leverage the hooks API for side effects. This provides a very elegant way to compose timing, delays, etc, that are quite hard to follow in plugin code and use utilities in the core library, cancelAllKeyboardResponses and clearAllTimeouts come to mind. I also think it is very easy to deal with configuration this way, instead of having to manually building the html string. The html string building method also makes it difficult to reason about the code imo. The complex logic involved in configuration the DOM elements can be dealt with using `useLayoutEffect()`.

**Comment by jodeleeuw** [comment]
> I don't fully understand jsPsych's timeline node system, in fact that was one of the motivations for me to write it this way so that the Javascript stack takes care of the nesting. I assume you cannot dynamically add trials anywhere between the timeline, which this API would allow you to do, so maybe they are incompatible?

One of my goals for a future version (probably v8.0) is to rewrite the backend of timeline management so that inserting trials dynamically is possible. This would be a moment to create a system that is more compatible with generators or async functions.

> I think the current declarative system is better for simple experiments i.e. the user does not need to care too much about looping and randomization. However I think when you have logic before and after trials and different behavior as a result of it, then the declarative API becomes complex very fast. While this approach might be slightly more complex to start, I think the composability it provides will be easier as they continue to use it. A lot of my experiments in math cognition / categorization would benefit from this API.

Certainly there are going to be designs that are easier to express in one version or the other. That's a great reason to have an alternative mode of constructing timelines. It may also depend on how the experimenter prefers to think about the experiment. Some folks may like expressing things in a declarative mode and others may like the generative mode. My hope is that either mode would allow writing any experiment.

> It's global as in there's only one, maybe singleton is a better choice of word. I am thinking of 'viewing' plugins, that display stimuli, and 'response' plugins, that deal with user input. Both plugins need a display_element, but deal with them differently. That way the user can compose between the type of response they'd like and the stimuli they'd like to show.

Got it, that makes sense. This would be a design philosophy change, but an interesting one to consider. It may be that moving some of the generic displays (image, audio, video, html) and generic responses (key, button, slider, text input) to a utility package would enable a plugin that can compose these different ideas together through its parameters, e.g.:

```js
var trial = {
	type: 'generic-plugin',
	stimulus: jsPsych.stimulus.image({ ... }),
	response: jsPsych.stimulus.button({ ... })
}
```

Changing things so that more than one plugin can be active at a time is a more radical change because it affects timeline design, data collection, and everything that depends on these systems.

> You can make it so that the users who don't care about the build chain don't need to, just provide them a transpiled bundle to use, kinda like how you can do so with React (without JSX of course). Those who want to use the library using npm can do so using the es module version. Given React's license, one can bundle it along with the library code, making it seamless for the user to use.

> I think the primary advantage to using React would be to leverage the hooks API for side effects. This provides a very elegant way to compose timing, delays, etc, that are quite hard to follow in plugin code and use utilities in the core library, cancelAllKeyboardResponses and clearAllTimeouts come to mind. I also think it is very easy to deal with configuration this way, instead of having to manually building the html string. The html string building method also makes it difficult to reason about the code imo. The complex logic involved in configuration the DOM elements can be dealt with using useLayoutEffect().

I'm still queasy about adding a dependency here. I think the open design flexibility of jsPsych's plugin system is a feature. Adding structure to this then requires people writing plugins to think about the code in a particular way. For folks who know React it would be great. For those who don't there's a whole new system to learn. 

That said, I'm in total agreement that *my implementation* of many plugins is based around a particular style that may be difficult to follow (or may be easy, depending on how one likes to think about it). A lack of enforced structure certainly makes it harder to follow what other people's code is doing. 

Some of the specifics that you mention, like `cancelAllKeyboardResponses` are things that can be fixed in the implementation of particular plugins or perhaps through some global modifications to the setup and tear down of plugins, making these bits of code unnecessary.  I'm all for finding ways to improve how plugins are managed, as long as their flexibility is maintained.

**Reply by vijaymarupudi** [reply]
> One of my goals for a future version (probably v8.0) is to rewrite the backend of timeline management so that inserting trials dynamically is possible. This would be a moment to create a system that is more compatible with generators or async functions.

Oh sweet, I'll keep experimenting with it then, so that I can get an idea of the issues at hand. I'm all for having multiple ways of expressing the experiment!

> Changing things so that more than one plugin can be active at a time is a more radical change because it affects timeline design, data collection, and everything that depends on these systems.

Indeed, and maybe this is a change that might never happen. I was considering this because people seem to ask for this sometimes and that answer is typically modify the plugin. If they could position the display elements, and decide what should be on each display element, then a lot of the configuration limitations would be dealt with.

Regarding React, I do think the plugin system should be independent of it, I kinda like the elegance of just providing the display element and trial settings and returning the data. I was suggesting using React in the plugins themselves when it is useful (ReactDOM.render() being in the plugin file on the display_element, not in the core itself). This might not necessarily be a change that jsPsych needs (the current plugins work fine and wonderfully), but something I was considering along with the other suggestions.

I'll try to improve the concept as I find time. I greatly value your feedback, and any tips and advice you might have in the needs of researchers or abotu jsPsych in general!
